staging/src/k8s.io/kubectl/pkg/cmd/set/set_image.go

```diff
@@ -119,6 +120,7 @@ func NewCmdImage(f cmdutil.Factory, streams genericiooptions.IOStreams) *cobra.C
 		Short:                 i18n.T("Update the image of a pod template"),
 		Long:                  imageLong,
 		Example:               imageExample,
+		ValidArgsFunction:     completion.SetImageCompletionFunc(f),
```

Comment:
I reviewed the context extensively and thanks for spending time and effort on this. I think, we don't need new `SetImageCompletionFunc` to achieve the completion of resources. 
I think, this code would work;
```go
ValidArgsFunction:     completion.SpecifiedResourceTypeAndNameNoRepeatCompletionFunc(f, validArgs),
```
and `validArgs`;
```go
validArgs := []string{"daemonsets", "deployments", "pods", "cronjobs", "replicasets", "replicationcontrollers", "statefulsets"}
```
and the changes in completion won't be needed?

Reply:
Something similar in rollout status https://github.com/kubernetes/kubernetes/blob/a5575425b039bf7c15dfaa9a7acf257fdc4fde3f/staging/src/k8s.io/kubectl/pkg/cmd/rollout/rollout_status.go#L99-L107

Reply:
Thanks for resolving this issue. I'll take a look at that function and test it out.
IDK how but i think i need to add you to the commit as well.

Reply:
It's close but it doesn't do the job as we wanted. it doesn't suggest container names it stops at pods.
The result on my machine:
`kubectl set image pods prometheus-operator-7d5c68dc4-tv7jv` and another TAB would not do anything else.
Let me see if i can change my function to a pattern like `SpecifiedResourceTypeAndNameNoRepeatCompletionFunc` and try again.

Reply:
Completion of containers would overcomplicate things. Users get used to complete pods (or other resource types) whereas there is no other command that has container completion.